lock-api/src/main/java/com/palantir/lock/client/LeasingTimelockClient.java

```diff
@@ -0,0 +1,154 @@
+import com.palantir.lock.v2.LockRequest;
+import com.palantir.lock.v2.LockResponse;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.StartAtlasDbTransactionResponse;
+import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionRequest;
+import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionResponse;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.lock.v2.WaitForLocksRequest;
+import com.palantir.lock.v2.WaitForLocksResponse;
+import com.palantir.timestamp.TimestampRange;
+
+public final class LeasingTimelockClient implements TimelockService {
+    private TimelockServerInterface delegate;
+    private LockLeaseManager lockLeaseManager;
+
+    private LeasingTimelockClient(TimelockServerInterface timelockService, LockLeaseManager lockLeaseManager) {
+        this.delegate = timelockService;
+        this.lockLeaseManager = lockLeaseManager;
+    }
+
+    public static LeasingTimelockClient create(TimelockServerInterface timelockServerInterface) {
+        return new LeasingTimelockClient(timelockServerInterface, LockLeaseManager.create());
+    }
+
+    @Override
+    public long getFreshTimestamp() {
+        return delegate.getFreshTimestamp();
+    }
+
+    @Override
+    public TimestampRange getFreshTimestamps(int numTimestampsRequested) {
+        return delegate.getFreshTimestamps(numTimestampsRequested);
+    }
+
+    @Override
+    public LockImmutableTimestampResponse lockImmutableTimestamp(IdentifiedTimeLockRequest request) {
+        return delegate.lockImmutableTimestamp(request);
+    }
+
+    @Override
+    public StartAtlasDbTransactionResponse startAtlasDbTransaction(IdentifiedTimeLockRequest request) {
+        return delegate.startAtlasDbTransaction(request);
+    }
+
+    @Override
+    public StartIdentifiedAtlasDbTransactionResponse startIdentifiedAtlasDbTransaction(
+            StartIdentifiedAtlasDbTransactionRequest request) {
+        long startTime = System.nanoTime();
+        LeasableStartIdentifiedAtlasDbTransactionResponse leasableResponse =
+                delegate.leasableStartIdentifiedAtlasDbTransaction(request);
+
+        StartIdentifiedAtlasDbTransactionResponse response = leasableResponse.getStartTransactionResponse();
+        Optional<Duration> leasePeriod = leasableResponse.getLeasePeriod();
+
+        updateLockLeases(response.immutableTimestamp().getLock(), startTime, leasePeriod);
+
+        return response;
+    }
+
+    @Override
+    public long getImmutableTimestamp() {
+        return delegate.getImmutableTimestamp();
+    }
+
+    @Override
+    public LockResponse lock(LockRequest request) {
+        long startTime = System.nanoTime();
+        LeasableLockResponse leasableResponse = delegate.leasableLock(request);
+
+        LockResponse lockResponse = leasableResponse.getLockResponse();
+        Optional<Duration> leasePeriod = leasableResponse.getLeasePeriod();
+
+        if (lockResponse.wasSuccessful()) {
```

Comment:
you are right... we will need to keep both the links from real token to generated one and generated to real one. That would at least require one more weakHashMap, I believe. There may be some tricks to avoid that though, let's sync on that offline.